File: packages/server/src/database/models/Account.ts
import { Column, Entity, OneToMany, PrimaryGeneratedColumn } from "typeorm";
import World from "./World";

@Entity()
export default class Account {
  @PrimaryGeneratedColumn("uuid")
  id!: string;

  // account details
  @Column({ nullable: false, unique: true, length: 20 })
  username!: string;

  @Column({ nullable: false, unique: true })
  email!: string;

  @Column({ nullable: false, unique: false, length: 64 })
  password!: string;

  // owned things in game

  @Column()
  maxEnergy!: number;

  /**
   * This is the amount of energy the user has, at a certain point in time (`timeEnergyWasAmount`).
   */
  @Column()
  lastEnergyAmount!: number;

  /**
   * The amount of milliseconds since the Unix epoch, that points to the point in time of which lastEnergyAmount was set.
   */
  @Column({ type: "bigint" })
  timeEnergyWasAtAmount!: string;

  /**
   * Describes how many milliseconds must go by for a single energy to be regenerated.
   */
  @Column()
  energyRegenerationRateMs!: number;

  get currentEnergy(): number {
    // if they have more than the maximum amount of energy, don't bother counting it.
    if (this.lastEnergyAmount >= this.maxEnergy) {
      return this.lastEnergyAmount;
    }

    const millisecondsSinceUnixEpoch = Date.now();
    const millisecondsEnergyHasBeenRegenerating = millisecondsSinceUnixEpoch - parseInt(this.timeEnergyWasAtAmount);
    const amountOfRegeneratedEnergyPrecise = millisecondsEnergyHasBeenRegenerating / this.energyRegenerationRateMs;
    const amountOfRegeneratedEnergy = Math.trunc(amountOfRegeneratedEnergyPrecise); // | 0 would also work here

    // cap the energy to maxEnergy
    return Math.min(this.lastEnergyAmount + amountOfRegeneratedEnergy, this.maxEnergy);
  }

  set currentEnergy(energy: number) {
    this.lastEnergyAmount = energy;

    const millisecondsSinceUnixEpoch = Date.now();
    this.timeEnergyWasAtAmount = "" + millisecondsSinceUnixEpoch;
  }

  @OneToMany(() => World, (world) => world.owner)
  worlds!: World[];
}

File: SFGServer/SFGServer.DAL/Models/Account.cs
namespace SFGServer.DAL.Models;

public class Account
{
    public Guid Id { get; set; }
    public string Username { get; set; } = null!;
    public string Email { get; set; } = null!;
    public string Password { get; set; } = null!;
    public int MaxEnergy { get; set; }
    public int LastEnergyAmount { get; set; }
    public long TimeEnergyWasAtAmount { get; set; }
    public int EnergyRegenerationRateMs { get; set; }

    public virtual ICollection<ShopItem> ShopItems { get; set; } = null!;
    public virtual ICollection<World> Worlds { get; set; } = null!;

    public uint GetEnergyAt(DateTime time)
    {
        var millisecondsSinceUnixEpoch = (long)time.Subtract(DateTime.UnixEpoch).TotalMilliseconds;
        var millisecondsEnergyHasBeenRegenerating = millisecondsSinceUnixEpoch - TimeEnergyWasAtAmount;
        var amountOfRegeneratedEnergyPrecise = (double)millisecondsEnergyHasBeenRegenerating / EnergyRegenerationRateMs;
        var amountOfRegeneratedEnergy = (int)Math.Floor(amountOfRegeneratedEnergyPrecise);
        var energyAtTime = Math.Min(amountOfRegeneratedEnergy, MaxEnergy);

        if (energyAtTime < 0)
        {
            throw new InvalidOperationException("The current energy of the player is negative somehow!");
        }

        return (uint)energyAtTime;
    }

    public void SetEnergyAt(DateTime time, uint energyAmount)
    {
        var millisecondsSinceUnixEpoch = (long)time.Subtract(DateTime.UnixEpoch).TotalMilliseconds;
        var millisecondsEnergyHasBeenRegenerating = millisecondsSinceUnixEpoch - TimeEnergyWasAtAmount;
        var amountOfRegeneratedEnergyPrecise = Math.Min(MaxEnergy, (double)millisecondsEnergyHasBeenRegenerating / EnergyRegenerationRateMs);
        var exactRemains = amountOfRegeneratedEnergyPrecise % 1.0d;
        var timeSpentGeneratingPartialEnergy = exactRemains * EnergyRegenerationRateMs;

        var nowWithRemains = millisecondsSinceUnixEpoch - timeSpentGeneratingPartialEnergy;

        if (energyAmount > int.MaxValue)
        {
            throw new InvalidOperationException("Cannot have more than int.MaxValue energy!");
        }

        LastEnergyAmount = (int)energyAmount;
        TimeEnergyWasAtAmount = (long)nowWithRemains;
    }
}
